oxo.py

```diff
@@ -1,26 +1,101 @@
             If the square is not empty, leave it as-is and return False. """
-        raise NotImplementedError("TODO: implement set_square")
+
+        if self.board[x, y] == 0:
+            self.board[x, y] = mark
+            return True
+        else:
+            return False
 
     def is_board_full(self):
         """ If there are still empty squares on the board, return False.
             If there are no empty squares, return True. """
-        raise NotImplementedError("TODO: implement is_board_full")
 
-    def get_winner(self):
+        return False if 0 in self.board.values() else True
+
+    def get_winner(self, row_length=3):
         """ If a player has three in a row, return 1 or 2 depending on which player.
             Otherwise, return 0. """
-        raise NotImplementedError("TODO: implement get_winner")
+
+        for cell in self.board:
+
+            if self.board[cell] != 0:
+
+                # NOTE: DISPLAY IS FLIPPED DIAGONALLY (oof)
+
+                # For each cell, we will search in 4 directions:
+
+                # Horizontally (to the right)
+
+                if cell[1] + row_length <= self.width:
+
+                    for y in range(cell[1] + 1, cell[1] + row_length):
+
+                        if self.board[cell[0], y] != self.board[cell]:
+                            break
+                        elif y == cell[1] + row_length - 1:
+                            return self.board[cell]
+
+                # Diagonally (top left to bottom right)
+
+                if cell[1] + row_length <= self.width \
+                        and cell[0] + row_length <= self.height:
+
+                    y = cell[1] + 1
+
+                    for x in range(cell[0] + 1, cell[0] + row_length):
+
+                        if self.board[x, y] != self.board[cell]:
+                            break
+                        elif x == cell[0] + row_length - 1:
+                            return self.board[cell]
+                        y += 1
+
+                # Diagonally (bottom left to top right)
+
+                if cell[1] + row_length <= self.width \
+                        and cell[0] - row_length + 1 >= 0:
+
+                    y = cell[1]
+
+                    for x in range(cell[0] - 1, cell[0] - row_length, -1):
+
+                        y += 1
+
+                        if self.board[x, y] != self.board[cell]:
+                            break
+                        elif x == cell[0] - row_length + 1:
+                            return self.board[cell]
+
+                # Vertically (downwards)
+
+                if cell[0] + row_length <= self.height:
+
+                    for x in range(cell[0] + 1, cell[0] + row_length):
+
+                        if self.board[x, cell[1]] != self.board[cell]:
+                            break
+                        elif x == cell[0] + row_length - 1:
+                            return self.board[cell]
```

Comment:
These 4 checks are all quite similar -- could they have been combined somehow to cut down on code repetition?